### PR TITLE
Log WAF policy context information

### PR DESCRIPTION
**Pull request by vkrivopalov**
## Description

With the introduction of per-site/per-URI support in WAF, we are going to have multiple WAF policies assigned to different contexts (global, listeners and locations) within a single gateway.
In order to be able to filter out messages for a specific policy and context, we need to log the following details:

- WAF policy identifier
- WAF scope (global, per-listener, per-location)
- WAF scope name

This fix adds corresponding functionality to ModSecurity.

## Testing

Tested locally with two different WAF policies and checked WAF log file with `tail -f`:
```
...
{
    "resourceId": "dummyResourceId", 
    "operationName": "ApplicationGatewayFirewall", 
    "category": "ApplicationGatewayFirewallLog", 
    "properties": {
        "instanceId": "dummyInstanceId", 
        "clientIp": "127.0.0.1", 
        "clientPort": "", 
        "requestUri": "/", 
        "ruleSetType": "", 
        "ruleSetVersion": "", 
        "ruleId": "980130", 
        "message": "Mandatory rule. Cannot be disabled. Inbound Anomaly Score Exceeded (Total Inbound Score: 5 - SQLI=5,XSS=0,RFI=0,LFI=0,RCE=0,PHPI=0,HTTP=0,SESS=0): SQL Injection Attack: SQL Tautology Detected.", 
        "action": "Blocked", 
        "site": "Global", 
        "details": {
            "message": "Warning. Operator GE matched 5 at TX:inbound_anomaly_score. ", 
            "data": "", 
            "file": "rules/RESPONSE-980-CORRELATION.conf", 
            "line": "73"
        }, 
        "hostname": "localhost:8080", 
        "transactionId": "zcAcAcAcAcAcWcA1AVAcAcAc", 
        "policyId": "secondPolicy", 
        "policyScope": "location", 
        "policyScopeName": "Two"
    }
}
{
    "resourceId": "dummyResourceId", 
    "operationName": "ApplicationGatewayFirewall", 
    "category": "ApplicationGatewayFirewallLog", 
    "properties": {
        "instanceId": "dummyInstanceId", 
        "clientIp": "127.0.0.1", 
        "clientPort": "", 
        "requestUri": "/", 
        "ruleSetType": "OWASP_CRS", 
        "ruleSetVersion": "3.0.0", 
        "ruleId": "942130", 
        "message": "SQL Injection Attack: SQL Tautology Detected.", 
        "action": "Matched", 
        "site": "Global", 
        "details": {
            "message": "Warning. Pattern match \"(?i:([\\s'\"`\\(\\)]*?)([\\d\\w]++)([\\s'\"`\\(\\)]*?)(?:(?:=|<=>|r?like|sounds\\s+like|regexp)([\\s'\"`\\(\\)]*?)\\2|(?:!=|<=|>=|<>|<|>|\\^|is\\s+not|not\\s+like|not\\s+regexp)([\\s'\"`\\(\\)]*?)(?!\\2)([\\d\\w]+)))\" at ARGS:svs .... ", 
            "data": "Matched Data: 1=1 found within ARGS:svs: 1=1", 
            "file": "rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf", 
            "line": "554"
        }, 
        "hostname": "localhost", 
        "transactionId": "GcAcAaDcYUAMGcAcgcABAcAc", 
        "policyId": "FirstPolicy", 
        "policyScope": "listener", 
        "policyScopeName": "One"
    }
}

...
```

Also tested that empty values are inserted when corresponding directives are missing from the ModSecurity configuration file.

### CODE REVIEW

File: apache2/apache2_config.c [hunk 1]
```diff
@@ -125,6 +125,11 @@ void *create_directory_config(apr_pool_t *mp, char *path)
     dcfg->stream_inbody_inspection = NOT_SET;
     dcfg->stream_outbody_inspection = NOT_SET;
 
+    /* WAF policy identification information */
+    dcfg->waf_policy_id = NOT_SET_P;
+    dcfg->waf_policy_scope = NOT_SET_P;
```
Comment (meetug) [inline apache2/apache2_config.c:130]: What is the difference between policy scope and site? In other words, can site be repurposed?

Comment (vkrivopalov) [reply apache2/apache2_config.c:130]: What do you mean by 'site'? Is it some other config directive?

Comment (meetug) [reply apache2/apache2_config.c:130]: No, an attribute in the log you pasted in the description.

Comment (vkrivopalov) [reply apache2/apache2_config.c:130]: Ah, I see. I am not sure as to what exactly this `site` attribute is used for in today's logs.
@azhao155 do you know what it means and what it is used for?

Comment (vkrivopalov) [reply apache2/apache2_config.c:130]: Looking at the code, it seems we always pass the value `0` for `site` which results into the `Global` string. Not really sure if it is needed at all, perhaps we can remove it altogether but we first need someone who knows why it has been added to provide us with some background.

Comment (azhao155) [reply apache2/apache2_config.c:130]: I am not sure the purpose of the site, but it's in v1 log, i think maybe for the css backward compatibility, we always keep it, may be need to check with @allanrbo

Comment (allanrbo) [reply apache2/apache2_config.c:130]: "Site" is there because we originally planned to do multi site (what you are doing now) shortly after releasing, but didn't succeed. It wasn't thought through properly.

File: apache2/apache2_config.c [hunk 2]
```diff
@@ -3448,6 +3539,30 @@ const command_rec module_directives[] = {
         "On or Off"
     ),
 
+    AP_INIT_TAKE1 (
```
Comment (allanrbo) [inline apache2/apache2_config.c:3542]: so this will make the config parameter work in apache, right? What about Nginx? Could we skip adding the apache support and only do it for nginx? Or is apache support a prereq for nginx support?

File: apache2/apache2_config.c (hunk 1 above)
Comment (vkrivopalov) [reply apache2/apache2_config.c:130]: I see. Technically we could re-use the "site" parameter without breaking backward compatibility as we'd still use "Global" for the per-gateway WAF policy, however, I personally find "scope" to better convey the meaning here (the scope to which the given WAF policy applies - global, listener or location). What do you folks think?

File: apache2/apache2_config.c (hunk 2 above)
Comment (vkrivopalov) [reply apache2/apache2_config.c:3542]: This will be supported through the regular ModSecurity configuration file (modsecurity.conf). Currently all ModSec directives, including those added earlier specifically for WAF such as SecWafResourceId, go there and I think it should stay this way. The only ModSec directives exposed to Nginx configuration are ModSecEnabled and ModSecConfig which implies that whatever configuration options are there are all put into the file pointed to by ModSecConfig. I'd prefer to keep it that way.

File: apache2/apache2_config.c (hunk 1 above)
Comment (vkrivopalov) [reply apache2/apache2_config.c:130]: I have added a comment explaining that 'site' field is deprecated. Let's not remove it for the time being to avoid unexpected regressions at customers' side where they may have logs parsing rely on it.

**Review by allanrbo — APPROVED** [review]
I'm fine with the code change. Whether this is the right functionality to add I guess is still an ongoing discussion in the thread with Amit.
